svf-llvm/lib/ObjTypeInference.cpp

```diff
@@ -590,69 +592,77 @@ u32_t ObjTypeInference::objTyToNumFields(const Type *objTy)
 
 
 /*!
- * get or infer the class names of thisptr
+ * get or infer the class names of thisptr; starting from :param:`thisPtr`, will walk backwards to find
+ * all potential sources for the class name. Valid sources include global or stack variables, heap allocations,
+ * or C++ dynamic casts/constructors/destructors.
+ * If the source site is a global/stack/heap variable, find the corresponding constructor/destructor to
+ * extract the class' name from (since the type of the variable is not reliable but the demangled name is)
  * @param thisPtr
- * @return
+ * @return a set of all possible type names that :param:`thisPtr` could point to
  */
 Set<std::string> &ObjTypeInference::inferThisPtrClsName(const Value *thisPtr)
 {
     auto it = _thisPtrClassNames.find(thisPtr);
     if (it != _thisPtrClassNames.end()) return it->second;
 
     Set<std::string> names;
-    auto insertClassNames = [&names](Set<std::string> &classNames)
+
+    // Lambda for checking a function is a valid name source & extracting a class name from it
+    auto addNamesFromFunc = [&names](const Function *func) -> void
     {
-        names.insert(classNames.begin(), classNames.end());
+        ABORT_IFNOT(isClsNameSource(func), "Func is invalid class name source: " + dumpValueAndDbgInfo(func));
+        for (auto name : extractClsNamesFromFunc(func)) names.insert(name);
```

Comment:
using`for (const auto& name : extractClsNamesFromFunc(func))` can avoid additional copies of string. It seems that `for (auto name : extractClsNamesFromFunc(func))` does not involve copy elision?

Reply:
Changed